### ISSUE TITLE
Native support for web sockets

### ISSUE THREAD

**Issue by AndreasHald** [sveltejs/kit#1491]
We've been attempting to migrate our app from sapper to sveltekit and ran into the problem of web sockets, in sapper we could simply attach the WebSocket server directly to the express server and be done with it, but as per Rich Harris' comments [here](https://github.com/sveltejs/kit/issues/334#issuecomment-804987028) and here [here](https://github.com/sveltejs/kit/issues/334#issuecomment-806289525) in sveltekit we don't want to expose direct access to the middlwares thus promoting the node adapter over the serverless ones.

However seeing as some of the serverless providers are starting to support web sockets ([cloudflare](https://blog.cloudflare.com/introducing-websockets-in-workers/)) ([begin](https://blog.begin.com/introducing-architect-5-0-fully-serverless-websockets-20bb97673c4e)) ([AWS Lambda](https://aws.amazon.com/blogs/compute/announcing-websocket-apis-in-amazon-api-gateway/)) should sveltekit provide a way to handle these connections directly? 

A native approach to this could also potentially provide a tool to solve this [issue](https://github.com/sveltejs/kit/issues/87) and [this](https://github.com/sveltejs/kit/issues/887)

**Describe the solution you'd like**
An interesting solution could be to expose functions in endpoints in the same way we handle http requests, and then let the adapters handle the rest

```routes/api/index.ts```
```js
// Handle post requests
export const post: RequestHandler = async (request) => {....}
// Handle web socket requests
export const ws: WebSocketRequestHandler = async (request) => {....}
```

A concern with this approach could be mixing protocols, within an endpoint could be confusing. Another could be if this becomes too complicated since as far as I know the initial [handshake](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#the_websocket_handshake) to start a WebSocket connection is handled over http, would this then be handled directly by sveltekit which could be considered too much "magic" since it would have to intercept a get request that would then not make it to the get handler in the endpoint. Or should we let the user implement the get endpoint and return the WebSocket pair and upgrade header.

**Describe alternatives you've considered**
The way we go about it today is a small script that manipulates the build output of sveltekit as such however this does not work in dev mode which is a pain point.
```js
const instance = createServer({ render: app.render }).listen(PORT, HOST, (err) => {
	if (err) {
		console.log('error', err);
	} else {
		console.log(`Listening on port ${PORT}`);
	}
});
app.server.installSubscriptionHandlers(instance);
export { instance };
```

Is this a feature that you feel would suit sveltekit? are there any considerations I've overlooked?

**Comment by Gregcop1** [comment]
@AndreasHald Have you ever heard of [Mercure](https://mercure.rocks/)?

**Comment by AndreasHald** [comment]
> @AndreasHald Have you ever heard of [Mercure](https://mercure.rocks/)?

Can't say i had. The idea looks interesting, but afaik sveltekit doesn't yet support SSE yet either.

We are pretty invested in Apollo Graphql server / client, and i don't think they have a transport layer for mercury, but correct me if i'm wrong.

**Comment by dunglas** [comment]
Hi! Mercure creator here.

> sveltekit doesn't yet support SSE yet either

Basically, as long as you delegate to a Mercure hub (a server implementing the protocol), you don't have to deal with SSE directly. The app just have to do a `POST` request to the hub to broadcast an update to all connected clients.

Supporting this part of the protocol perfectly fits with serverless environments because you can delegate the long running connection (that is by nature incompatible with serverless) to a dedicated component, even if the serverless service itself is not compatible with SSE, Websockets and the like.

> We are pretty invested in Apollo Graphql server / client, and i don't think they have a transport layer for mercury

I don't know Apollo very well. I know that API Platform has [native support for GraphQL subscriptions through Mercure](https://api-platform.com/docs/core/graphql/#subscriptions), and AFAIU there is no need for client-side code. Using the native `EventSource` JS class looks enough. But maybe that a better integration with Apollo could be created?

Anyway, I would love to see support for the protocol land in SvelteKit. Let me know if I can help in any way to make this happen (I can try to create a transport for Apollo for instance).

**Comment by cayter** [comment]
I had the same requirement and initially got some dirty workarounds to make it work with just SvelteKit but it's not maintainable. I have come to understand the direction SvelteKit is being steered towards(i.e. serverless rendering) that makes it quite tough to be fully compatible with the traditional NodeJS SSR + Vite itself would still lead us to a lot of bundling issues for server-side libraries. After playing with SvelteKit for a few weeks, I decided to build a set of toolkit on top of SvelteKit and brings the best possible monolith DX. For more details, please find my comment [here](https://github.com/sveltejs/kit/issues/887#issuecomment-877572191).

Though I haven't added:

- `export const ws` to support websocket
- `export const graphql` to support GraphQL
- `export const grpc` to support GRPC (haven't dived deep into this yet but will be very much needed for successful startups who wants to slowly move some monolith pieces out to another service)

but yes, the above are all planned! I'm just excited to see this post here as I was wondering the same why we can't support `export const ws` and `export const graphql`! Good job.

**Comment by AndreasHald** [comment]
@cayter sounds interesting, i think we will migrate to use SSE when [1568](https://github.com/sveltejs/kit/issues/1563) is resolved, as i understand there is a PR ready.

How would you go about implementing export const graphql.

Svelte kit supports the graphql standard very well out og the box. Because graphql specifies a single endpont just use a single file /graphql/index.js that supports the spec

We use apollo server without any issues.

> I had the same requirement and initially got some dirty workarounds to make it work with just SvelteKit but it's not maintainable. I have come to understand the direction SvelteKit is being steered towards(i.e. serverless rendering) that makes it quite tough to be fully compatible with the traditional NodeJS SSR + Vite itself would still lead us to a lot of bundling issues for server-side libraries. After playing with SvelteKit for a few weeks, I decided to build a set of toolkit on top of SvelteKit and bring out the best possible monolith DX. For more details, please find my comment [here](https://github.com/sveltejs/kit/issues/887#issuecomment-877572191).
> 
> 
> 
> Though I haven't added:
> 
> 
> 
> - `export const ws` to support websocket
> 
> - `export const graphql` to support GraphQL
> 
> - `export const grpc` to support GRPC (haven't dive deep into this yet but will be very much needed for successful startups who wants to slowly move some monolith pieces out to another service)
> 
> 
> 
> but yes, the above are all planned! I'm just excited to see this post here as I was wondering the same why we can't support `export const ws` and `export const graphql`! Good job.

**Comment by cayter** [comment]
> Svelte kit supports the graphql standard very well out og the box. Because graphql specifies a single endpont just use a single file /graphql/index.js that supports the spec We use apollo server without any issues.

I probably need to give it a try again, I was bumping into issues where Vite was bundling some of the server-side dependencies in the client bundle and broke the page rendering. But yeah I was thinking of support graphql (with subscription) easily via `export const graphql` though I haven't really come to this step yet, will share more once I'm onto it.

**Comment by benmccann** [comment]
I think we will be exposing a way to add middleware in `adapter-node` (https://github.com/sveltejs/kit/pull/2051), but if anyone wants to take a stab at adding more generic support that also support the serverless platforms, that still sounds valuable

**Comment by Xananax** [comment]
So is there some guideline regarding how to integrate sockets with a svelte-kit app? As far as I understand, the path of least resistance at the moment is to use `adapter-node`, and add a file that uses the built svelte site:

```js
import { assetsMiddleware, prerenderedMiddleware, kitMiddleware } from './build/middlewares.js';
import polka from 'polka';
import Socket from 'socket.io';

const { PORT=3000 } = process.env;

const app = polka();

app.use(assetsMiddleware, prerenderedMiddleware, kitMiddleware);

const { server } = polka().listen(PORT, () => console.log(`> Running on localhost:${PORT}`))

const io = Socket(server);
io.on('connection', socket => {
  console.log('connection')
})
```

I didn't try this yet, but it seems to be the recommended path. It bothers me somewhat because it isn't integrated in the build system (won't have access to the same state, won't be using typescript unless I set up a separate build, ...).

But assuming I go with that, but I am not sure how access to all the session information that I have in my svelte-kit app. I am also not very clear on changing the state of the socket server based on url calls. Should I send two requests from each URL in my svelte-app (one normal http call and one to the websocket server?) Should I add a middleware to `adapter-node`?

**Comment by matths** [comment]
> I didn't try this yet, but it seems to be the recommended path. It bothers me somewhat because it isn't integrated in the build system

@Xananax, well I did. And actually it is easy to integrate with the build process. You can specify your custom server as `entryPoint` in your `svelte.config.js` file. Your can read about that here: https://github.com/sveltejs/kit/tree/master/packages/adapter-node#entryPoint

An example:
* Put your custom server at your [YOUR_PROJECT_DIR]/server/index.js right next to [YOUR_PROJECT_DIR]/src
* Do not forget to add polka to your dev dependencies and to run `npm install`again
* Add `entryPoint: 'entryPoint: 'server/index.js','` in your adapter option within the kit.adapter section of your `svelte.config.js`
* run `npm build` and see the magic happen.

**Comment by gg187on** [comment]
hello, as part of school exam, we need to contribute to some OpenSource project, and bcs I like Svelte, I would like to contribute and solve this issue. I have idea of creating endpoint function like ws() and graphql(). I'm not sure about how complex this issue is, so I would like if someone would navigate me though. So I'm thinking about using socket.io and graphql-js libs for implementing this, would it be ok? What API would You suggest for these endpoint methods? And where should I start?

**Comment by benmccann** [comment]
I don't think that we'd want to provide built-in endpoints in that manner. I would really like to provide the feedback to your teacher that this assignment is not fair to open source maintainers. I welcome people getting involved in open source, but would encourage them to do so because of their own experience with the project. I've had this happen numerous times where people need to check a box because of a school assignment and it almost never results in code that we can accept to the project and instead just places more burden on us. This isn't your fault at all and I'm happy you're interested in Svelte so I hate to write something discouraging. But I do wish this trend of making us do more work for other people's school assignments would die. Your teacher should be responsible for reviewing all code changes and not making us do it. I am simply not interested in being a TA and will reject any contributions to the project without review if they are for a school assignment because it is not a scalable use of our time to do teaching and review of school assignments. That's a teacher's job and they should stop trying to foist their job upon others who didn't volunteer for such duties

**Comment by gg187on** [comment]
Ok, I understand, even tho, I'm already using Svelte for some time, and used it in few projects (even my bachelor project, and few school's exams), so I'm not new to Svelte. But I agree that this is not fair to open source maintainers. I will send Your message to profesor,...

**Comment by bfanger** [comment]
### Workaround
Inject socket.io _(or another websockets implementation)_ into the vite **DEV** server via the svelte.config: 
```js
import { Server } from "socket.io";
export default {
 kit: {
    vite: {
      plugins: [
        {
          name: "multiplayer",
          configureServer(server) {
              const io = new Server(server.httpServer));
              // do websocket stuff
            });
          },
        },
      ],
      ...
```
https://github.com/bfanger/multiplayer-dice-game/blob/main/svelte.config.js 

For **PROD** use the node-adapter to generate the middleware and which you'll include in a custom server:
https://github.com/bfanger/multiplayer-dice-game/blob/main/server.js

**Comment by glics** [comment]
Using adapter-node as middleware along with a websocket server outside of sveltekit means that, if for example you are using TypeScript and .ts database models (i.e.: my current structure is lib/db/models/[...].ts, fetched from endpoints), you'll need a lot of extra steps to make these usable. In my case this makes it not worth it to even support real time updates :/

**Comment by sjmueller** [comment]
@Rich-Harris any guidance on the priority to support websockets in ‘22? We have a realtime api based on Elixir and Phoenix, and the story around websockets is the only aspect giving us pause with adopting sveltekit.

**Comment by benmccann** [comment]
> It bothers me somewhat because it isn't integrated in the build system (won't have access to the same environment variables and import.meta goodies, won't be using typescript unless I set up a separate build, ...).

This should now be addressed since https://github.com/sveltejs/kit/pull/2931. Previously, `adapter-node` had a second bundling step using esbuild. Now Vite does everything

**Comment by bluepuma77** [comment]
@benmccann It would be great if we could get a quick description (or link to it) how we can use WebSocket with SvelteKit. 

Specifically for those web developers not so much involved with the inner workings of SvelteKit and Vite.

**Comment by kubikdanon** [comment]
+1 for WebSockets. I love SvelteKit and wanted to build a real-time game on it, but lack of support requires me to separate my back end and front end at this point.

**Comment by superboss224** [comment]
As of now, Cloudflare Workers, Deno Deploy, AWS are supporting WebSockets
There is gonna be more and more WebSocket support on serverless platform, having some native and first-class WebSocket support is something we truly need to make immersive experiences that can go realtime

**Comment by kevin192291** [comment]
Any updates on this?

**Comment by VirgileD** [comment]
This really is the last piece that's missing in SK to finally replace all my other stacks.
Hope it will be there soon :-)

**Comment by cayter** [comment]
Just FYI, for those who are not running on serverless technology, you can rely on [this](https://github.com/sveltejs/kit/tree/master/packages/adapter-node#custom-server) to setup your own HTTP server alongside the websocket server. I had successfully setup a Fastify server + [tRPC](https://trpc.io/) with subscriptions.

**Comment by kevin192291** [comment]
I am running my own setup (not serverless) I would finally decided to run a node instance as my web-socket server. It is all well and good, but it would be really nice to have this built-in, as it would greatly reduce the maintenance and code complexity. This is currently my ONLY sveltekit ask!!!

**Comment by jdgamble555** [comment]
Cloud Run supports Websockets as well, which is a serverless technology... this could change the game in frameworks. I predict this will be a thing one day, just hopefully sveltekit before nextjs. This is so important for modern app development and would change the game.

J

**Comment by CropWatchDevelopment** [comment]
As scary as it sounds, I think that this may be a make or break feature for SvelteKit VS Nextjs

**Comment by cayter** [comment]
It seems `solid-start` included the initial implementation for [websocket](https://github.com/solidjs/solid-start/blob/main/examples/with-websocket/src/routes/room/%5Bid%5D.tsx#L4), probably SvelteKit can borrow some ideas from it too.

Disclaimer: I'm not sure if it only plays well with `node` adapter or the serverless platform like netlify/cloudflare worker as well.

**Comment by zachjensz** [comment]
Hopefully we'll have native websockets very soon but anyone looking for a solution in the meantime should check out JoyOfCode's [video](https://www.youtube.com/watch?v=mAcKzdW5fR8) or [article](https://joyofcode.xyz/using-websockets-with-sveltekit) on a workaround.

**Comment by Eudritch** [comment]
When will SvelteKit have support for WebSocket?

**Comment by JumpyLionnn** [comment]
Are there any news about the websocket support? I need it for my project and when i tried to separate the client and the server it became very messy.

**Comment by josias-r** [comment]
Just keep in mind that "native support" is [impossible](https://vercel.com/guides/do-vercel-serverless-functions-support-websocket-connections) for any serverless deployment, such as [vercel](https://vercel.com/).

This is because this type of deployment does not have a permanently running server that could send WS data to it's clients or keep the connection up in the first place. Using the svelte `node` adapter means opting-out of serverless deployment.

Once you realize that, maybe "native support" is not quite as intreaguing as you think. For my project, I opted for a mono-repo architecture like this. so I can keep all the serverless benefits:
```yml
/app # the sveltekit app, w/ adapter-auto
/server # a simple express server, w/ socket.io
/shared # shared code. I use it for TypeScript types and shared utility methods, etc. With some tsconfig tinkering you can get your project so you can import your shared code like so: `import xy from '$shared/myCode'`, for both the svelte and the server
```

Then I simply deploy my svelte app to vercel and my node express server w/ socket.io to some conventional host.

**Comment by jdgamble555** [comment]
> Just keep in mind that "native support" is [impossible](https://vercel.com/guides/do-vercel-serverless-functions-support-websocket-connections) (afaik) for any serverless deployment, such as [vercel](https://vercel.com/).

It is definitely **NOT** impossible for serverless technologies to support a version of web sockets:
- [Cloud Run](https://cloud.google.com/run/docs/triggering/websockets)
- [Cloudflare](https://support.cloudflare.com/hc/en-us/articles/200169466-Using-Cloudflare-with-WebSockets) - (which is what Vercel uses under-the-hood for edge functions)


Even if you were right, which is not entirely the whole picture, Vercel could theoretically create a different deployment option like the `/api` folder, say `/socket`, which would deploy differently than the rest of the app. It is already doing this with edge functions, static files, and regular aws lambdas.

The thing is, this would need to start with Vercel -- given that Vercel supports SvelteKit -- then trickle down to SvelteKit -- and other adapters and services would be modified later to support it.  

J

**Comment by Brl23** [comment]
Considering that this is a feature that it seems a lot of people are interested in, I don't see why we can't begin to thrash out options for the API. It means that when it comes to someone making a pull request, everyone is on the same page, and in agreement about the specifics (or at least roughly), which will surely speed up the process.

**Comment by carlosV2** [comment]
Hi. If someone is interested, I just finished a [little package](https://www.npmjs.com/package/@carlosv2/adapter-node-ws) to embed WebSockets into SvelteKit in an straightforward way for both DEV and PROD environments. It is, by no means, a bullet proof solution but a summary of techniques used by other people all hacked into a handy package.

I will be using it for one of my projects but, unfortunately, it might not work for everybody.

Ultimately, my intention is to deprecated it as soon as SvelteKit finally comes with an official way of doing WebSockets.

**Comment by suhaildawood** [comment]
Inspired by the implementation in [trpc-sveltekit](https://github.com/icflorescu/trpc-sveltekit), I published a reference architecture for anyone who's interested in integrating a WebSocket server in their SvelteKit application. Workarounds previously discussed here involve either modifying SvelteKit internals or bundling WebSocket middleware via a separate step build step, both of which have various drawbacks. This example is as close to first-class support for WebSockets as you could get: it works on both development and production, allows you to define startup functionality and logic (like authentication) in `hooks.server.ts` and by attaching the WebSocket server instance to `event.locals` within the `handler` function, all server-side endpoints have access to the WebSocket server instance. Instructions and a more detailed write-up can be found in the repository: https://github.com/suhaildawood/SvelteKit-integrated-WebSocket

**Comment by darkuranium** [comment]
I'm finding myself also needing this.

In terms of design, I kind of like the approach proposed here: https://joyofcode.xyz/using-websockets-with-sveltekit

Something along the lines of (in a page's `+server.{js,ts}`):

```typescript
/*
 * The return type is something along the lines of:
 *   type WebSocketResponse =
 *       { websocket?: { connected(ws: WebSocket): unknown; ... } } //< accepted
 *       | Response; //< rejected
 * ... or similar. For example, `websocket` could simply be part of `new Response(..., { websocket: { ... } })`,
 * or it could be a `new WebSocketResponse({ ... })`.
 */
export const WS: WebSocketRequestHandler = (event) => {
    if(!allowWSConnection(event))
        throw error(401, 'Unauthorized');
    return {
        connected(ws) { ... },
        ...
    };
};
```

This is just some brainstorming, not really a concrete proposal.
Another alternative would be for `GET` to handle accepting/rejecting the websocket (but in this case, I would recommend providing an easy way to check if the client is requesting a WebSocket connection, without having to parse `Connection` and/or `Upgrade`!). Then, `WS` could simply accept a `WebSocket`-typed parameter, with the newly-created websocket (plus perhaps the request that created it?).

The major advantage of this general approach would be the ability to reuse the existing routing to handle websockets, making the entire thing nicely consistent.

**Comment by ttmx** [comment]
Just here to bump the issue, working with websockets feels very clumsy right now, this really is the only feature left that I feel is needed in sveltekit.

**Comment by Trackhe** [comment]
Not thinking about quic ^^' @ttmx

**Comment by Dennis4720** [comment]
I'm missing this as well.

**Comment by nicholasvperry** [comment]
Take a loot at this post. I am not sure if this is any help. [Tutorial](https://joyofcode.xyz/using-websockets-with-sveltekit)

**Comment by dmoebius** [comment]
FYI: In #927 I outlined a solution which initializes a websocket handler (socket.io in this case) into the Polka http server generated by node-adapter which uses the build system _and_ works in all three modes dev, preview and production.

**Comment by lukaszpolowczyk** [comment]
Is there an option, as a THOUSAND solution, to allow sw functions to be placed, in `export const config`?
And it would only be a function for nodejs, it would only work with nodejs adapter.

**Comment by faulander** [comment]
Can someone point me to a working solution to implement *only* the client into SvelteKit? Background: DIRECTUS now also has websocket support and currently i am polling the database. Having a subscription would be awesome here, but i seem to be unable to get the client implementation to work.

**Comment by Trackhe** [comment]
> Can someone point me to a working solution to implement *only* the client into SvelteKit? Background: DIRECTUS now also has websocket support and currently i am polling the database. Having a subscription would be awesome here, but i seem to be unable to get the client implementation to work.

I guess this https://joyofcode.xyz/using-websockets-with-sveltekit is actually the best what you can find.

**Comment by llmaboi** [comment]
> Can someone point me to a working solution to implement _only_ the client into SvelteKit? Background: DIRECTUS now also has websocket support and currently i am polling the database. Having a subscription would be awesome here, but i seem to be unable to get the client implementation to work.

I've made a few different repositories with svelte-kit and some web socket logic. Please ignore my bad ts-ignores.
https://github.com/llmaboi/planning-poker-sveltekit-trpc/blob/main/src/routes/room/%5BroomId%5D/%5BdisplayId%5D/%2Bpage.svelte#L46-L70

**Comment by TomDo1234** [comment]
> Can someone point me to a working solution to implement _only_ the client into SvelteKit? Background: DIRECTUS now also has websocket support and currently i am polling the database. Having a subscription would be awesome here, but i seem to be unable to get the client implementation to work.

Just install socket io and import that in the client side script tags

**Comment by faulander** [comment]
> 

No, that doesn't work. socket.io only supports socket.io

**Comment by alanxp** [comment]
> https://github.com/llmaboi/planning-poker-sveltekit-trpc/blob/main/src/routes/room/%5BroomId%5D/%5BdisplayId%5D/%2Bpage.svelte#L46-L70

even throught thats the best option for now, its not made with typescript and that adds a whole new level of complexity

**Comment by llmaboi** [comment]
> > https://github.com/llmaboi/planning-poker-sveltekit-trpc/blob/main/src/routes/room/%5BroomId%5D/%5BdisplayId%5D/%2Bpage.svelte#L46-L70
> 
> even throught thats the best option for now, its not made with typescript and that adds a whole new level of complexity

I agree I wish there was a more type safe way I thought of to do this, maybe in future I can, but at least for now there is a workable way. I could probably get it to be more type safe, but would require additional work.

**Comment by luke-hagar-sp** [comment]
Web Socket support is my ONLY wishlist item for SvelteKit

**Comment by PhatDump** [comment]
I yearn for WS support in SvelteKit

**Comment by alanxp** [comment]
why the team doesnt seem to put interest on websockets? they were the future, now they arent but still they are very important

**Comment by benmccann** [comment]
I understand there's a lot of interest in this, but please don't comment expressing interest. Instead use the thumbs up on the issue. Commenting with expressions of interest makes it harder to find the comments discussing how to implement this and we do need to be able to easily find those details to come up with a design for this. I've hidden all comments that are not relevant to the implementation of the feature.

The core team won't be implementing this in the near future as we're currently focused on a very exciting set of changes in Svelte 5 as detailed in our roadmap: https://svelte.dev/roadmap. However, we will periodically switch back to adding major new features to SvelteKit in the future. You can read more about our process here: https://github.com/sveltejs/svelte/blob/master/CONTRIBUTING.md#our-process.

In team meantime, we do still welcome folks chiming in with technical details that may help this get implemented. Examples that we can follow, lists of which platforms/adapters can and cannot support the feature, etc. are all helpful. We're also continuing to review PRs from the community if anyone would like to take a stab at this.

**Comment by PhatDump** [comment]
This seems like a very well-done implementation that provides control of the WebSocket server to SvelteKit's server logic:
https://github.com/suhaildawood/SvelteKit-integrated-WebSocket/tree/main

Surprised nobody has linked to this , yet.

**Comment by Loizzus** [comment]
Bun has [a socket server](https://bun.sh/docs/api/websockets) built in. Thoughts on whether or not it could be a [viable alternative to node.js](https://bun.sh/guides/ecosystem/sveltekit) and solve our socket issue?

**Comment by ADIOP55550** [comment]
> Bun has [a socket server](https://bun.sh/docs/api/websockets) built in. Thoughts on whether or not it could be a [viable alternative to node.js](https://bun.sh/guides/ecosystem/sveltekit) and solve our socket issue?

Also they suggest [sveltekit-adapter-bun](https://github.com/bun-community/sveltekit-adapter-bun), which has websocket support built-in via their [Exposed Websocket Server](https://github.com/bun-community/sveltekit-adapter-bun#exposed-websocket-server). 
I'm planning to use it soon so I will see if it works.

**Comment by PimTournaye** [comment]
Any updates on this, with SvelteKit 2?

**Comment by benmccann** [comment]
We just merged https://github.com/sveltejs/kit/pull/11649, which adds a `supports` feature as part of the adapter API to allow adapters to indicate whether they support an optional feature. This could potentially be leveraged for future web socket support allowing us to indicate that `adapter-node` and `adapter-cloudflare` support websockets while the other adapters in this repo don't.

**Comment by ashwin31** [comment]
> We just merged #11649, which adds a `supports` feature as part of the adapter API to allow adapters to indicate whether they support an optional feature. This could potentially be leveraged for future web socket support allowing us to indicate that `adapter-node` and `adapter-cloudflare` support websockets while the other adapters in this repo don't.

then we can add socket.io server to sveltekit like how we do with express js ?
i am running express js + socket.io server separately because sveltekit is not supporting socket.io server.

**Comment by craigcosmo** [comment]
need socket. Please support

**Comment by aolose** [comment]
About using websocket with adapter-node, I have an example to share [here](https://github.com/aolose/sk-node-ws-demo) and it works well. I wrote a [vite plugin](https://github.com/aolose/vite-sveltekit-node-ws) with monkey patches to hack the `node:http` package to get the httpServer. Then you can use your favorite websocket framework normally. 

As for adapter-cloudflare you can refer to [this](https://github.com/aolose/sk-cf-ws-demo). 

In fact, the need for websockets represents the user's desire to handle raw requests before entering sveltekit. It is true that incorrect handling can  break sveltekit, but the user will be responsible for this.

I hope Sveltekit can provide a hook  like `import { onMount } from 'svelte'`

It could look like below, and it will be handled before the request is processed by sveltekit .

```js
import {onServerMount} from 'svelteKit';
import {WebSocketServer} from 'ws';

onServerMount((httpServerInstance, preRequestHandle) => {
    // for adapter-node
    const ws = new WebSocketServer(httpServerInstance)
    ws.on('connection',(ws)=>{

    })

    // such as cloudflare
    preRequestHandle((req, res, next) => {
        const upgradeHeader = request.headers.get('Upgrade');
        if (!upgradeHeader || upgradeHeader !== 'websocket')return next()
        const webSocketPair = new WebSocketPair();
        const client = webSocketPair[0],
            server = webSocketPair[1];
        server.accept()
        server.on('connection',(ws)=>{

        })
        res.status(101).set('webSocket',client)
    })
})
````

**Comment by jacobbogers** [comment]
Hi I just wanted to chime in

Websockets where specifically designed to piggy back on existing http request response  functionality.

There are 2 questions if we want to implement this:

1.  how would a websocket connection be exposed to kit. I think making it look like we currently use api routes would be a good idea and very developer friendly. aka a Specific handler name.
2. what needs to change in kit code to make above possible

Firstly:
What is a websocket request? what happens when someone types `new WebSocket('/chat')` for example.
In that case we see just a normal `GET` request with specific "upgrade headers" on the server side.

_Note that the parameter in `WebSocket` is an actual url/path , this could be the route path to `routes/chat/+server.ts`_
 
So how can we implement websockets keeping the feel of api routes?

I see 2 possibilities:

1.  kit detects GET request (+ upgrade headers), creates a context for the duration of the websocket lifespan and whenever a data is sent over the line it calls  `wsEventData(data)`, in `routes/chat/+server.ts`. the return value would be the response to said data. The difficulty in this approach is how would the server send data to the client if the server initiates the send...   I think therefor 2. (next point) is more appropreate

2. kit detects GET request (+ upgrade headers), and calls the function `UPGRADE` in `routes/chat/+server.ts` **after** successfull network  upgrade.



```javascript
// called AFTER kit has upgraded the GET request to websocket protocol and ws "open" event is issued
export function UPGRADE(
     event: RequestEvent, // original event for the ws upgrade
     ws: WebSocket  // 
 ) {
     
  
    ws.onmessage = e => {...}
    ws.onerror = e => {...}
    ws.onclose = e => {...}
         
    // WebSocket has  `reqId` property 
    // ws is stored in a global object so if you want to send something you need to fetch the websocket from a global map
}
```

**Comment by ADIOP55550** [comment]
For me it would make sense to have it a little more explicit. I am thinking of something like

```typescript
// /src/routes/.../+server.ts
export const websocket: WebSocketHandler = async ({wss}: {wss:WebSocketServer}) => {
   wss.on("connection", (socket) => {
      socket.on("open", ()=>{/*...*/});
      socket.on("message", ()=>{/*...*/});
      socket.on("close", ()=>{/*...*/});
      socket.on("error", ()=>{/*...*/});

      // It would also allow for server initiated conversation
      socket.emit(/*...*/);
   }
   // Broadcasting to all clients would be possible
   wss.broadcast(/*...*/);
};
```

OR [the bun way](https://github.com/bun-community/sveltekit-adapter-bun?tab=readme-ov-file#exposed-websocket-server):

```typescript
export const handleWebsocket = {
    message: (ws, msg) => {
        ws.send(msg);
    },
    //optionally, doesnt have to be set - can be used to deny websocket
    upgrade: (req, upgrade) => {
        upgrade(req, {data:1});
    }
}
```

**Comment by MahmoodKhalil57** [comment]
[jacobbogers](https://github.com/jacobbogers)
> Hi I just wanted to chime in
> 
> Websockets where specifically designed to piggy back on existing http request response functionality.

any implementation should take this in mind

**Comment by LukeHagar** [comment]
> Hi I just wanted to chime in
> 
> Websockets where specifically designed to piggy back on existing http request response functionality.
> 
> There are 2 questions if we want to implement this:
> 
> 1. how would a websocket connection be exposed to kit. I think making it look like we currently use api routes would be a good idea and very developer friendly. aka a Specific handler name.
> 2. what needs to change in kit code to make above possible
> 
> Firstly: What is a websocket request? what happens when someone types `new WebSocket('/chat')` for example. In that case we see just a normal `GET` request with specific "upgrade headers" on the server side.
> 
> _Note that the parameter in `WebSocket` is an actual url/path , this could be the route path to `routes/chat/+server.ts`_
> 
> So how can we implement websockets keeping the feel of api routes?
> 
> I see 2 possibilities:
> 
> 1. kit detects GET request (+ upgrade headers), creates a context for the duration of the websocket lifespan and whenever a data is sent over the line it calls  `wsEventData(data)`, in `routes/chat/+server.ts`. the return value would be the response to said data. The difficulty in this approach is how would the server send data to the client if the server initiates the send...   I think therefor 2. (next point) is more appropreate
> 2. kit detects GET request (+ upgrade headers), and calls the function `UPGRADE` in `routes/chat/+server.ts` **after** successfull network  upgrade.
> 
> ```js
> // called AFTER kit has upgraded the GET request to websocket protocol and ws "open" event is issued
> export function UPGRADE(
>      event: RequestEvent, // original event for the ws upgrade
>      ws: WebSocket  // 
>  ) {
>      
>   
>     ws.onmessage = e => {...}
>     ws.onerror = e => {...}
>     ws.onclose = e => {...}
>          
>     // WebSocket has  `reqId` property 
>     // ws is stored in a global object so if you want to send something you need to fetch the websocket from a global map
> }
> ```

Assuming that the request that would trigger the `UPGRADE` function here is still processed by hooks (server, and client), for handling things like auth guards, or populating relevant data. I think this is a pretty elegant solution.

We also want to ensure solutions for handling sets of connections, or accessing a specific connection is well supported.
```ts
// Broadcast message to a specific user's clients
function broadcastMessage(userID: string, message: any) {
  if (clients[userID]) {
    clients[userID].forEach((socketID) => {
      io.to(socketID).emit("message", message);
    });
  }
}
```

This is still the biggest/only thing missing for me using SK day to day.

**Comment by Padi2312** [comment]
+1

**Comment by xcaptain** [comment]
Any progress on this? need a websocket endpoint in sveltekit app

**Comment by benmccann** [comment]
Work is on n progress. We're waiting on https://github.com/unjs/crossws/pull/91

**Comment by pi0** [comment]
throwable Responses and shared context supported in [crossws@0.3.2](https://github.com/unjs/crossws/releases/tag/v0.3.2).

I haven't followed this thread but if could do any help please don't hesitate to ask! (you can reach me by `@pi0` in discord)

**Comment by ivo22dev** [comment]
Hello! I'm trying to use cloudflare web socket. The only problem i can't resolve is that at the client side when doing:

```
await fetch('', {
			method: 'GET',
			headers: {
				Upgrade: 'websocket'
			}
		});
```
i recieve the error:
`TypeError: Worker tried to return a WebSocket in a response to a request which did not contain the header "Upgrade: websocket".`


I seems that sveltekit isn't sendind the header even if i explicit code it.

Any clue to resolve it? Thanks!

**Comment by benmccann** [comment]
The web socket feature is now ready for testing! We'd love your feedback.

You can install the version of SvelteKit from the PR using pkg.pr.new and the PR number. E.g. if you're using `adapter-node` you would do:
```
npm i https://pkg.pr.new/sveltejs/kit/@sveltejs/kit@12973
npm i https://pkg.pr.new/sveltejs/kit/@sveltejs/adapter-node@12973
```

Please see the PR for docs: https://github.com/sveltejs/kit/pull/12973

**Comment by pi0** [comment]
Made an issue in crossws: https://github.com/unjs/crossws/issues/138

**Comment by nerg4l** [comment]
Thanks. I will continue the discussion here. I couldn't figure out how to pass the node response to `handleRequest`. I can access the response through the `platform` variable but not the response. I might need to open a separate issue for this.

Also, it seems `platform` is not populated with `req` in dev mode. I'm not sure if this should happen when I run `vite dev` but it would help a lot during development.

**Comment by nerg4l** [comment]
I experimented with the current PR and was able to replace Socket.IO rooms with crossws topics. My next problem is I cannot access `RequestEvent` object from any of the hooks, not even the upgrade hook.

**Comment by nerg4l** [comment]
@ciscoheat, at the moment, you can use a [custom node server](https://redislabs.atlassian.net/browse/RED-134225) to use Socket.IO **or** use an adapter with Socket.IO support. This issue tracks adding native WS support, but I wasn't able to make Socket.IO work with the current proposed implementation. As you can see in my previous message, I just started to use crossws instead.

**Comment by alanxp** [comment]
Testing is very difficult at this point

> > Please ignore this question if you think it's irrelevant or off-topic. I was wondering how someone could integrate Socket.IO with this native web socket implementation. I know that will only work with the node-adapter, and I found https://socket.io/how-to/use-with-nuxt#hook-the-socketio-server for Nuxt, but I got a bit stuck when I needed to pass the node request/response object.
> 
> [@nerg4l](https://github.com/nerg4l) to answer your question about how to use Socket.io, I expect you could follow their Nuxt docs as an example since Nuxt also uses `crossws`. Hopefully after we officially add web sockets they can add a SvelteKit page to their docs too 😄
> 
> https://socket.io/how-to/use-with-nuxt#hook-the-socketio-server

Testing is very difficult at this point

**Comment by alanxp** [comment]
> The web socket feature is now ready for testing! We'd love your feedback.
> 
> You can install the version of SvelteKit from the PR using pkg.pr.new and the PR number. E.g. if you're using `adapter-node` you would do:
> 
> ```
> npm i https://pkg.pr.new/sveltejs/kit/@sveltejs/kit@12973
> npm i https://pkg.pr.new/sveltejs/kit/@sveltejs/adapter-node@12973
> ```
> 
> Please see the PR for docs: [#12973](https://github.com/sveltejs/kit/pull/12973)

Testing is very difficult at this point

every time i test i get error

WebSocket connection to 'ws://localhost:5173/ws' failed: 

with:
		primary_socket = new WebSocket('/ws');

**Comment by codeit-ninja** [comment]
I am testing the WebSocket implementation too, works well thus far.

I am wondering tho, how to access the socket from other places in the app? Lets assume I want to send a message to all peers that listen on /routes/ws/+server.ts from /routes/api/updater/+server.ts, how do I do that?

Currently I am thinking of connecting to the WebSocket using the NodeJS ws package, then send message to the peers through a connection. Seems a bit cumbersome tho, should we not expose the socket from somewhere? That we can import the connection in other routes? Just like locals for example?

**Comment by thomaswiradikusuma** [comment]
@cliffordkleinsr does it work with Cloudflare Workers?

**Comment by dev-itzarun** [comment]
> @cliffordkleinsr does it work with Cloudflare Workers?

Does it works?

**Comment by dev-itzarun** [comment]
> The web socket feature is now ready for testing! We'd love your feedback.
> 
> You can install the version of SvelteKit from the PR using pkg.pr.new and the PR number. E.g. if you're using `adapter-node` you would do:
> ```
> npm i https://pkg.pr.new/sveltejs/kit/@sveltejs/kit@12973
> npm i https://pkg.pr.new/sveltejs/kit/@sveltejs/adapter-node@12973
> ```
> 
> Please see the PR for docs: https://github.com/sveltejs/kit/pull/12973


Can we get a Cloudflare workers support for this ?

**Comment by teemingc** [comment]
> > The web socket feature is now ready for testing! We'd love your feedback.
> > You can install the version of SvelteKit from the PR using pkg.pr.new and the PR number. E.g. if you're using `adapter-node` you would do:
> > ```
> > npm i https://pkg.pr.new/sveltejs/kit/@sveltejs/kit@12973
> > npm i https://pkg.pr.new/sveltejs/kit/@sveltejs/adapter-node@12973
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Please see the PR for docs: [#12973](https://github.com/sveltejs/kit/pull/12973)
> 
> Can we get a Cloudflare workers support for this ?

You should be able to install it by replacing the package name in the URL, so something like:
```sh
npm i https://pkg.pr.new/sveltejs/kit/@sveltejs/adapter-cloudflare@12973
```
Two things to note:
1. This doesn't currently work with Durable Objects so there's no coordination between different Workers.
2. We will be working on a new implementation so many things are subject to change.

**Comment by benmccann** [comment]
While we're waiting for native SvelteKit support, here's an example using `polka` and the `WebSocketServer` from the `ws` package. 

Vite config:
```
import { sveltekit } from '@sveltejs/kit/vite';
import { setupWebSocketServer } from './websocket.js';
import { type ViteDevServer, defineConfig } from 'vite';
import { WebSocketServer } from 'ws';

export default defineConfig({
	plugins: [
		sveltekit(),
		{
			name: 'vite-plugin-websocket',
			apply: 'serve',
			configureServer(server: ViteDevServer) {
				if (!server.httpServer) throw new Error('httpServer not available');
				const wss = new WebSocketServer({ port: 24678 });
				setupWebSocketServer(wss, server.httpServer);
			}
		}
	]
});
```

Custom entry point for use with `adapter-node` custom server:
```
import http from 'node:http';
import polka from 'polka';
import { WebSocketServer } from 'ws';
import { setupWebSocketServer } from './websocket.js';
import { handler } from './build/handler.js'; // the SvelteKit app

const server = http.createServer();

const app = polka({ server });

const wss = new WebSocketServer({ server });
setupWebSocketServer(wss, server);

app.use(handler);

app.listen(3000, (err) => {
	if (err) throw err;
	console.log('listening on port 3000');
});

```

websocket file:
```
export function setupWebSocketServer(wss, server) {
	wss.on('connection', (ws) => {
		let counter = 0;
		const interval = setInterval(() => {
			counter++;
			ws.send(JSON.stringify({ count: counter }));
		}, 1000);

		ws.on('close', () => {
			clearInterval(interval);
		});

		server.on('close', () => {
			wss.close();
		});
	});
}
```

`+page.svelte` file to call the websocket:
```
<script>
	import { onMount, onDestroy } from 'svelte';

	let count = $state(0);
	let ws;

	onMount(() => {
		const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
		ws = new WebSocket(
			`${protocol}//${import.meta.env.DEV ? 'localhost:24678' : window.location.host}`
		);

		ws.onmessage = (event) => {
			const data = JSON.parse(event.data);
			count = data.count;
		};

		ws.onerror = (error) => {
			console.error('WebSocket error:', error);
		};
	});

	onDestroy(() => {
		if (ws) {
			ws.close();
		}
	});
</script>

<div>
	<h1>WebSocket Counter</h1>
	<p>Count: {count}</p>
</div>
```

**Comment by AgarwalPragy** [comment]
@benmccann Unfortunately, that doesn't allow the websocket server to share code with the sveltekit server.
Svetelkit uses path aliases like `$lib` and loads env variables from `$env`, which don't work outside of sveltekit context.

I managed to set up code sharing via the following

`vite.config.ts` Vite plugin to run websocket server in sveltekit context during dev
```ts
import { sveltekit } from '@sveltejs/kit/vite'
import { defineConfig, type Plugin } from 'vite'
import { Server } from 'http'

const realtimeServerPlugin: Plugin = {
    name: 'realtimeServer',
    configureServer(server) {
        if (!server.httpServer || !(server.httpServer instanceof Server)) return

        server.httpServer.on('listening', () => {
            server.ssrLoadModule('src/lib/server/realtime/index.ts').then(module => {
                const { startRealtimeServer } = module as {
                    startRealtimeServer: (server: import('http').Server) => Promise<void>
                }

                startRealtimeServer(server.httpServer as Server).catch(err => {
                    server.config.logger.error('Failed to start real-time server:', err)
                    process.exit(1)
                })
            })
        })
    },
}


export default defineConfig({
    plugins: [sveltekit(), realtimeServerPlugin]
})
```

`server.js` Customer server for `adapter-node` that runs both websocket and sveltekit servers
```js
import express from 'express'
import { createServer } from 'http'
import { handler } from './build/handler.js'

const app = express()
const server = createServer(app)
const PORT = process.env.PORT || 3000

const startRealtimeServer = globalThis.startRealtimeServer

if (startRealtimeServer) {
    startRealtimeServer(server).catch(err => {
        console.error('Failed to start real-time server:', err)
        process.exit(1)
    })
} else {
    console.error('Real-time server boot function not found. The build may be broken.')
}

app.use(handler)

server.listen(PORT, () => { console.log(`Server listening on http://localhost:${PORT}`) })
```

`src/hooks.server.ts` Attach the websocket object to `globalThis`. Forces vite to bundle the websocket code in the final build output.
```ts
import type { Handle } from '@sveltejs/kit'
import { startRealtimeServer } from '$lib/server/realtime'

export const handle: Handle = async ({ event, resolve }) => {
    return resolve(event)
}

globalThis.startRealtimeServer = startRealtimeServer
```

The websocket server code can now be put inside `src/lib/server/realtime/*` and can share code (types, db schemas, repositories, services) with your sveltekit server code.


Wondering if this approach is okay, or whether there's some glaring issues with doing things like this. 😅

**Comment by anaksulung99** [comment]
I tried to make a simple configuration, Works perfectly in development and production using adapter-node or adapter-bun

[https://github.com/ketarketir/sveltekit-ws](https://github.com/ketarketir/sveltekit-ws)

**Comment by seantiz** [comment]
Any chance integration get easier with `WebTransport` API, now that it's supported by all browsers?

**Comment by aurorarissime** [comment]
Gentle bump :)
